### DISCUSSION TITLE
Can we avoid .nonempty() changing the inferred input type?

### DISCUSSION THREAD

**Discussion by Balinth** [colinhacks/zod#3040] [Q&A] [answered]
Take below example:

```ts
import { z } from 'zod';

const schema = z.number().array().nonempty();

// expected type to be number[], but instead is [number, ...number[]]
type inputType = z.input<typeof schema>;

// simplified for demo
const formValidation: (userInput: inputType) => boolean = (userInput) => {
  // ....
  return true;
};

// user input of either correct array or incorrect empty array...
const formData: number[] = Math.random() % 2 ? [2] : [];

// this should typecheck, but doesn't because the .nonemtpy() call pollutes the inferred input type.
const isValid = formValidation(formData);
```
Is this a bug? or intentional?

I can imagine zod being intentionally like this, as this behaviour is the one that utilises TS to the limit, but I would expect that this is the opposite of what the usual use case would demand.

Is there some elegant solution to this, that doesn't involve breaking typesafety with multiple schema definitions and/or forced casts especially when the array is actually in a nested zod object schema, so I can't easily just create a simple and a chained .nonemtpy variant?

**Comment by JacobWeisenburger** [comment, marked as answer]
TypeScript is doing what it's supposed to do here:
```ts
const schema = z.number().array().nonempty()

type NonEmptyArray = z.input<typeof schema>
// type NonEmptyArray = [number, ...number[]]

const foo = ( x: NonEmptyArray ) => { }
foo( [ 42 ] ) // ok
foo( [] )
//   ^^
// Argument of type '[]' is not assignable to parameter of type '[number, ...number[]]'.
//   Source has 0 element(s) but target requires 1.
```

If you truly want a compile time non empty array, then this is correct. But it seems you don't want your types to match your runtime values. Is that correct?

**Comment by Balinth** [comment]
Note that I am talking about `z.input` and not `z.infer`. I do want the end type to match the type described by the schema.

What I am surprised by is that `.nonemtpy()` doesn't behave like `.transform(...)` and modifies the inferred input type as well.

Since then I realized that I can easily achieve what I need with using `.refine(...).transform(...)` instead of `.nonempty()` and check for minimum number of elements and signifying that the output type is a not empty array that way, without forcing the input type to be a nonempty array.

**Reply by CiprianaM** [reply]
Wouldn't a more straight forward solution be 
`const schema = z.number().array().min(1);`
?

**Comment by tanjunior** [comment]
i found this https://github.com/causaly/zod-non-empty-array/
